### ISSUE TITLE
Add a warning when writing a Python file inside virtual environments 

### ISSUE THREAD

**Issue by luabud** [microsoft/pylance-release#4975]
This is similar to https://github.com/microsoft/pylance-release/issues/3550 but when users save their Python files inside a virtual environment. Apparently there are tutorials out there that are directing users to save their Python files inside venvs and the Python extension does not behave well like that. It'd be nice if we could show an error when we when users are trying to change and save a file, so users can either move their files outside of a virtual environment, or just know they might be making edits to the wrong file.

**Comment by erictraut** [comment]
How do you propose to determine when a user saves a file in a virtual environment? How does that look different from any other source file in a virtual environment?

**Comment by karthiknadig** [comment]
For a virtual environment at this location: `C:\GIT\project\.venv`. 
The python path for this would be:
Windows: `C:\GIT\project\.venv\scripts\python.exe`
Linux\mac: `C:\GIT\project\.venv\bin\python`

Following location are under the `venv`:
```
C:\GIT\project\.venv
C:\GIT\project\.venv\lib\site-packages
```
Adding files under `C:\GIT\project\.venv` should be treated as bad practice.

This is probably what we want to tell the users. From `venv` documentation: https://docs.python.org/3/library/venv.html
```
A virtual environment is (amongst other things):
- Used to contain a specific Python interpreter and software libraries and binaries which are needed to support a project (library or application). These are by default isolated from software in other virtual environments and Python interpreters and libraries installed in the operating system.
- Contained in a directory, conventionally either named venv or .venv in the project directory, or under a container directory for lots of virtual environments, such as ~/.virtualenvs.
- Not checked into source control systems such as Git.
- Considered as disposable – it should be simple to delete and recreate it from scratch. You don’t place any project code in the environment
- Not considered as movable or copyable – you just recreate the same environment in the target location.
```

**Comment by karthiknadig** [comment]
> How does that look different from any other source file in a virtual environment?

It does not. Editing the files inside the `venv` itself is the bad practice here. I think, if the user adds their project files, or edits a file that belongs to some third-party package, all should be treated the same way.

**Comment by erictraut** [comment]
> ... if the user adds their project files

What operation are you suggesting would trigger such an error? A save operation? I don't think language servers are told when a file is saved to disk.

> ... or edits a file

Are you suggesting that files opened from within a venv should be marked as read-only so they cannot be edited? Some users might find that annoying, so it would probably need to be a configuration option. Also, I'm not aware of a way for a language server to tell VS Code that certain files must be treated as read-only, but perhaps there's some mechanism I don't know about.

**Comment by luabud** [comment]
We wouldn't mark the files as read-only, the idea is to just show a warning that could be dismissed and disabled if the action is intentional. 

From what I understood when chatting with @rchiodo it sounded like it was possible to trigger such a warning on a file save request?

**Comment by rchiodo** [comment]
Yeah we would do this on the VS code client using its API. It wouldn't be in Pyright.

**Comment by heejaechang** [comment]
This issue has been fixed in prerelease version 2026.3.102, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20263102-4-September-2026-PreRelease)
